server/src/main/java/org/elasticsearch/repositories/IndexSnapshotsService.java

```diff
@@ -0,0 +1,172 @@
+                );
+            }
+        );
+
+        final Repository repository = getRepository(repositoryName);
+        if (repository == null) {
+            listener.onFailure(new RepositoryMissingException(repositoryName));
+            return;
+        }
+
+        final String indexName = shardId.getIndexName();
+        StepListener<RepositoryData> repositoryDataStepListener = new StepListener<>();
+        StepListener<FetchShardSnapshotContext> snapshotInfoStepListener = new StepListener<>();
+
+        repositoryDataStepListener.whenComplete(repositoryData -> {
+            if (repositoryData.hasIndex(indexName) == false) {
+                listener.onResponse(Optional.empty());
+                return;
+            }
+
+            final IndexId indexId = repositoryData.resolveIndexId(indexName);
+            final List<SnapshotId> indexSnapshots = repositoryData.getSnapshots(indexId);
+
+            final Optional<SnapshotId> latestSnapshotId = indexSnapshots.stream()
+                .map(snapshotId -> Tuple.tuple(snapshotId, repositoryData.getSnapshotDetails(snapshotId)))
+                .filter(s -> s.v2().getSnapshotState() != null && s.v2().getSnapshotState() == SnapshotState.SUCCESS)
+                .filter(s -> s.v2().getStartTimeMillis() != -1 && s.v2().getEndTimeMillis() != -1)
+                .max(START_TIME_COMPARATOR)
+                .map(Tuple::v1);
+
+            if (latestSnapshotId.isEmpty()) {
+                // It's possible that some of the backups were taken before 7.14 and they were successful backups, but they don't
+                // have the start/end date populated in RepositoryData. We could fetch all the backups and find out if there is
+                // a valid candidate, but for simplicity we just consider that we couldn't find any valid snapshot. Existing
+                // snapshots start/end timestamps should appear in the RepositoryData eventually.
+                listener.onResponse(Optional.empty());
+                return;
+            }
+
+            final SnapshotId snapshotId = latestSnapshotId.get();
+            repository.getSnapshotInfo(
+                snapshotId,
+                snapshotInfoStepListener.map(
+                    snapshotInfo -> new FetchShardSnapshotContext(repository, repositoryData, indexId, shardId, snapshotInfo)
+                )
+            );
+        }, listener::onFailure);
+
+        snapshotInfoStepListener.whenComplete(fetchSnapshotContext -> {
+            final SnapshotInfo snapshotInfo = fetchSnapshotContext.snapshotInfo;
+
+            if (snapshotInfo == null || snapshotInfo.state() != SnapshotState.SUCCESS) {
+                // We couldn't find a valid candidate
+                listener.onResponse(Optional.empty());
+                return;
+            }
+
+            // TODO: This is executed in SNAPSHOT_META thread pool and we perform a few blocking operations here, should we execute this
+            // step somewhere else?
```

Comment:
I think this is ok, the `SNAPSHOT_META` threadpool is an appropriate place to block on a read of some snapshot metadata. Wouldn't hurt to document with an assertion that we're on this threadpool tho.

Reply:
👍 Addressed in e8fd26b493f9fcc0c05a483926286bd6821fe2e7